File: kevoree-core/org.kevoree.kompare/src/main/scala/org/kevoree/kompare/scheduling/SchedulingWithTopologicalOrderAlgo.scala
package org.kevoree.kompare.scheduling

/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.traverse.TopologicalOrderIterator
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.kevoreeAdaptation.AdaptationPrimitive
import org.kevoree._

class SchedulingWithTopologicalOrderAlgo {

  private val logger: Logger = LoggerFactory.getLogger(classOf[SchedulingWithTopologicalOrderAlgo])

  def schedule (commands: List[AdaptationPrimitive], start: Boolean): List[AdaptationPrimitive] = {
    if (commands.size > 1) {
      val graph: DefaultDirectedGraph[AdaptationPrimitive, (AdaptationPrimitive, AdaptationPrimitive)] = buildGraph(commands,
                                                                                                            start)
      val topologicAlgorithm: TopologicalOrderIterator[AdaptationPrimitive, (AdaptationPrimitive, AdaptationPrimitive)] = new
          TopologicalOrderIterator(graph)

      var listCommands = List[AdaptationPrimitive]()
      while (topologicAlgorithm.hasNext) {
        listCommands = listCommands ++ List(topologicAlgorithm.next)
      }
      if (listCommands.size == commands.size) {
        return listCommands
      }
    }
    commands
  }

  /**
   * each command is a vertex and edges represent dependency order between commands
   */
  private def buildGraph (commands: List[AdaptationPrimitive],
    start: Boolean): DefaultDirectedGraph[AdaptationPrimitive, (AdaptationPrimitive, AdaptationPrimitive)] = {
    val graph = new
        DefaultDirectedGraph[AdaptationPrimitive, (AdaptationPrimitive, AdaptationPrimitive)](classOf[(AdaptationPrimitive, AdaptationPrimitive)])

    val map = lookForPotentialConstraints(commands)

    for (command <- commands) {
      for (command2 <- commands) {
        graph.addVertex(command2)
        if (!command.equals(command2)) {
          (map.get(command2.getRef.asInstanceOf[Instance])) match {
            case Some(cmdDep) => {
              if (cmdDep.contains(command.getRef.asInstanceOf[Instance])) {
                if (start) {
                  graph.addEdge(command2, command, (command2, command))
                } else {
                  graph.addEdge(command, command2, (command, command2))
                }
              }
            }
            case _ =>
          }
        }

      }
    }
    graph
  }

  /*
     * Return Map
     *
     * Instance Map key is a dependency of all List instances return by key
     *
     *
     * [i:Instance,li:List[Instance]]
     * li depends i
     *
     *
     * */
  private def lookForPotentialConstraints (
    commands: List[AdaptationPrimitive]): scala.collection.mutable.Map[Instance, java.util.List[Instance]] = {
    val instanceDependencies: scala.collection.mutable.Map[Instance, java.util.List[Instance]] = scala.collection
      .mutable.Map[Instance, java.util.List[Instance]]()

    var rootContainer: ContainerRoot = null
    val firstCommand = (commands(0)).getRef.asInstanceOf[Instance]
    firstCommand match {
      case c: Group => rootContainer = c.eContainer.asInstanceOf[ContainerRoot]
      case c: Channel => rootContainer = c.eContainer.asInstanceOf[ContainerRoot]
      case c: ComponentInstance => rootContainer = c.eContainer.asInstanceOf[KevoreeContainer].eContainer.asInstanceOf[ContainerRoot]
    }


    if (rootContainer.getMBindings != null) {
      val bindingIterator = rootContainer.getMBindings.iterator
      while (bindingIterator.hasNext) {
        val binding = bindingIterator.next
        for (command <- commands) {
          command.getRef.asInstanceOf[Instance] match {
            case instance: ComponentInstance => {
              // test if instance is the container of the port of the binding
              if (instance.equals(binding.getPort.eContainer)) {
                // test all provided port
                // the instance of the provided port must be stopped before those which are connected to him
                val pit = instance.getProvided.iterator
                while (pit.hasNext) {
                  val port = pit.next
                  if (binding.getPort.equals(port)) {
                    val newL = instanceDependencies.get(instance).getOrElse(new java.util.ArrayList[Instance]())
                    newL.add(binding.getHub)
                    instanceDependencies.update(instance, newL)
                  }
                }
                // test all required port
                // the instance wait stops of all of those which are connected to him
                val rit = instance.getRequired.iterator
                while (rit.hasNext) {
                  val port = rit.next
                  // !port.getPortTypeRef.getNoDependency is used to be sure that we need to know if the use of this port on the start or stop method of the component can introduce deadlock or not
                  if (binding.getPort.equals(port) &&
                    (port.getPortTypeRef.getNoDependency == null || !port.getPortTypeRef.getNoDependency)) {
                    val newL = instanceDependencies.get(binding.getHub).getOrElse(new java.util.ArrayList[Instance]())
                    newL.add(instance)
                    instanceDependencies.update(binding.getHub, newL)
                  }
                }
              }
            }
            case _ =>
          }
        }
      }
    }
    instanceDependencies
  }
}

File: kevoree-core/org.kevoree.kompare/src/main/java/org/kevoree/kompare/scheduling/SchedulingWithTopologicalOrderAlgo.kt
package org.kevoree.kompare.scheduling

import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.traverse.TopologicalOrderIterator
import org.kevoree.*
import java.util.ArrayList
import org.kevoreeadaptation.AdaptationPrimitive
import java.util.HashMap
import org.kevoreeadaptation.ParallelStep
import org.kevoree.kompare.StepBuilder
import org.kevoreeadaptation.KevoreeAdaptationFactory
import org.kevoree.kompare.JavaSePrimitive
import org.kevoree.framework.kaspects.PortAspect
import org.kevoree.modeling.api.KMFContainer

class SchedulingWithTopologicalOrderAlgo: StepBuilder {
    override var previousStep: ParallelStep? = null
    override var currentSteps: ParallelStep? = null
    override var adaptationModelFactory: KevoreeAdaptationFactory = org.kevoreeadaptation.impl.DefaultKevoreeAdaptationFactory()

    fun schedule(commands: List<AdaptationPrimitive>, start: Boolean): /*List<AdaptationPrimitive>*/ParallelStep? {
        clearSteps()
        nextStep()
        val firstStep = currentSteps
        if (commands.size > 1) {
            val graph = buildGraph(commands, start)
            val topologicAlgorithm = TopologicalOrderIterator(graph)

            while (topologicAlgorithm.hasNext()) {
                val element = topologicAlgorithm.next()
                val list = ArrayList<AdaptationPrimitive>()
                list.add(element)
                if (start) {
                    createNextStep(JavaSePrimitive.StartInstance, list)
                } else {
                    createNextStep(JavaSePrimitive.StopInstance, list)
                }
                nextStep()
            }
        } else if (!commands.isEmpty()) {
            if (start) {
                createNextStep(JavaSePrimitive.StartInstance, commands)
            } else {
                createNextStep(JavaSePrimitive.StopInstance, commands)
            }
        }
        return firstStep;
    }


    private class Assoc2<E, G>(_1: E, _2: G){
    }


    /**
     * each command is a vertex and edges represent dependency order between commands
     */
    private fun buildGraph(commands: List<AdaptationPrimitive>,
                           start: Boolean): DefaultDirectedGraph<AdaptationPrimitive, Assoc2<AdaptationPrimitive, AdaptationPrimitive>> {

        val clazz = Assoc2<AdaptationPrimitive, AdaptationPrimitive>(commands.get(0), commands.get(0)).javaClass

        val graph = DefaultDirectedGraph<AdaptationPrimitive, Assoc2<AdaptationPrimitive, AdaptationPrimitive>>(clazz)

        //        val map = lookForPotentialConstraints(commands)
        val map = lookForPotentialConstraints(commands)

        for (command in commands) {
            for (command2 in commands) {
                graph.addVertex(command2)
                if (!command.equals(command2)) {
                    val cmdDep = (map.get(command2.getRef() as Instance))
                    if(cmdDep != null){
                        if (cmdDep.contains(command.getRef() as Instance)) {
                            if (start) {
                                graph.addEdge(command2, command, Assoc2(command2, command))
                            } else {
                                graph.addEdge(command, command2, Assoc2(command, command2))
                            }
                        }
                    }
                }
            }
        }
        return graph
    }

    /*
       * Return Map
       *
       * Instance Map key is a dependency of all List instances return by key
       *
       *
       * [i:Instance,li:List[Instance]]
       * li depends i
       *
       *
       * */
    private fun lookForPotentialConstraints(commands: List<AdaptationPrimitive>): Map<Instance, List<Instance>> {
        val instanceDependencies: HashMap<Instance, MutableList<Instance>> = HashMap<Instance, MutableList<Instance>>()

        for (command in commands) {
            if(command.getRef() is ComponentInstance) {
                val component = command.getRef() as ComponentInstance
                var instancesDependencyForComponent: MutableList<Instance>? = instanceDependencies.get(component)
                if(instancesDependencyForComponent == null){
                    instancesDependencyForComponent = ArrayList<Instance>()
                }

                // Looking for channel that send message to the instance
                for (port in component.getProvided()) {
                    for (binding in port.getBindings()) {
                        // the channel is a dependency of the instance (must be start after and stop before the instance)
                        val channel = binding.getHub()!!
                        if (!instancesDependencyForComponent!!.contains(channel)) {
                            var instancesDependencyForChannel: MutableList<Instance>? = instanceDependencies.get(channel)
                            if(instancesDependencyForChannel == null){
                                instancesDependencyForChannel = ArrayList<Instance>()
                            }
                            for (bindingFromChannel in channel.getBindings()) {
                                val portFromBinding = bindingFromChannel.getPort()!!
                                // each required port connected to the channel is host by a component that is a dependency of the instance (must be start after and stop before the instance) and a dependency of the channel
                                if (PortAspect().isRequiredPort(bindingFromChannel.getPort()!!) && !portFromBinding.getPortTypeRef()!!.getNoDependency() && !instancesDependencyForChannel!!.contains(bindingFromChannel.getPort()!!.eContainer() as Instance)) {
//                                    System.out.println((portFromBinding.eContainer() as NamedElement).getName() + "\t" + portFromBinding.getPortTypeRef()!!.getName() + "\t" + portFromBinding.getPortTypeRef()!!.getNoDependency())
//                                    System.out.println(component.getName() + " -> " + channel.getName())
                                    instancesDependencyForComponent!!.add(channel)

//                                    System.out.println(component.getName() + " -> " + (bindingFromChannel.getPort()!!.eContainer() as NamedElement).getName())
//                                    System.out.println(channel.getName() + " -> " + (bindingFromChannel.getPort()!!.eContainer() as NamedElement).getName())
                                    instancesDependencyForComponent!!.add(portFromBinding.eContainer() as Instance)
                                    instancesDependencyForChannel!!.add(portFromBinding.eContainer() as Instance)
                                }
                            }
                            instanceDependencies.put(channel, instancesDependencyForChannel!!)
                        }
                    }
                }
                instanceDependencies.put(component, instancesDependencyForComponent!!)
            }
        }
        return instanceDependencies
    }

}
